src/dune_rules/install_rules.ml

```diff
@@ -130,42 +139,57 @@ end = struct
-                    if Path.Local.is_root parent then None
-                    else
-                      Some (Path.Local.explode parent |> String.concat ~sep:"/")
-                  in
-                  (subdir, Some (Path.Local.basename p))
-                | None ->
-                  let dst =
-                    Path.Build.basename source
-                    |> String.drop_suffix ~suffix:"-gen"
-                  in
-                  let sub_dir =
-                    let src_dir = Path.Build.parent_exn source in
-                    if Path.Build.equal src_dir lib_src_dir then None
-                    else
-                      Path.Build.local src_dir
-                      |> Path.Local.descendant
-                           ~of_:(Path.Build.local lib_src_dir)
-                      |> Option.map ~f:Path.Local.to_string
-                  in
-                  (sub_dir, dst)
-              in
-              let sub_dir =
-                match sub_dir with
-                | None -> lib_subdir
-                | Some subdir ->
-                  Some
-                    (match lib_subdir with
-                    | None -> subdir
-                    | Some lib_subdir -> Filename.concat lib_subdir subdir)
+                let sub_dir, dst =
+                  match Module.install_as m with
+                  | Some p ->
+                    let subdir =
+                      let parent = Path.Local.parent_exn p in
+                      if Path.Local.is_root parent then None
+                      else
+                        Some
+                          (Path.Local.explode parent |> String.concat ~sep:"/")
+                    in
+                    (subdir, Some (Path.Local.basename p))
+                  | None ->
+                    let dst =
+                      Path.Build.basename source
+                      |> String.drop_suffix ~suffix:"-gen"
+                    in
+                    let sub_dir =
+                      let src_dir = Path.Build.parent_exn source in
+                      if Path.Build.equal src_dir lib_src_dir then None
+                      else
+                        Path.Build.local src_dir
+                        |> Path.Local.descendant
+                             ~of_:(Path.Build.local lib_src_dir)
+                        |> Option.map ~f:Path.Local.to_string
+                    in
+                    (sub_dir, dst)
+                in
+                (in_sub_dir sub_dir, dst)
               in
               make_entry ?sub_dir Lib source ?dst))
     in
+    let* melange_runtime_deps =
+      let* expander = Super_context.expander sctx ~dir:lib_src_dir in
+      Melange_rules.eval_runtime_deps ~expander (snd lib.melange_runtime_deps)
+    in
+    let melange_runtime_entries =
+      Path.Set.to_list_map
+        ~f:(fun path ->
+          let path = Path.as_in_build_dir_exn path in
+          let sub_dir =
+            let src_dir = Path.Build.parent_exn path in
```

Comment:
Same questions: in which cases could it fail here? and are they reproducible in tests?